FreeTubeApp/FreeTube
# Add video batch selection mode

# Add video batch selection mode




## Pull Request Type


## Related issue
closes #2517 

## Description
Implements a Selection Mode toggle for selecting multiple videos at once and performing all the batch actions that you would on a singular video.

- Select button on the top nav bar that changes hover/click/(multi-)select/[(multi-)touch](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events) behavior to "select" a video with visible styling
- When "Select" mode is enabled, a "three dots" button appears on the top nav right next to it (disabled if no selection is made yet)
  - The dropdown labels become plural, if applicable, when 2 or more videos are selected (use i18n interpolation)
    - With exact # of affected videos per action, e.g., save 6 videos, download 8 videos.
- Only applies to videos (if other types of results appear, e.g., playlists, hashtags, or channels in search results, they cannot be selected in Select mode)
- Each video's three-dot, favorite, etc. behavior only affects that video (no change)
- Copying multiple links: by default, has a newline delimiter between each link
- Select Mode dropdown has a Save/Unsave option
- Select Mode dropdown has a "Clear selections" option
- When Select mode is enabled, Ctrl+A selects all videos in the Selection mode manner
- Selections are automatically cleared when Select Mode is toggled off
- Select Mode is toggled off on route change
- Users cannot open more than 15 links at one time

Incidental:
- Adds & implements previously missing "More Options" label in localization

## Video 
https://github.com/FreeTubeApp/FreeTube/assets/84899178/e095b903-0881-4f7b-9d71-647c9e14cc86


## Testing 
- Confirm Select Mode is switched to disabled & all videos are unselected when switching routes or toggling the button
- Confirm that selection styling is never missing on any selected video, & counts always match up
- Confirm that videos can be selected and manipulated properly in all places that videos can appear (search results, playlists, history, channel pages, recommended videos, community pages)
- Try coping multiple links
- Confirm all visible videos are selected on pressing Ctrl+A in select mode, and ensure that the whole page is not selected when select mode is disabled
- Select 15 videos and confirm that link-opening options are present and functional
- Select 16 or more videos and confirm that link-opening options are no longer present
- Examine selection styling in all/different themes
- Select an incredibly large number of videos in selection mode; then try favoriting/unfavoriting, adding to/removing from history, etc., and ensure that nothing is breaking (e.g., counts being accurate, whether all videos are actually being affected etc.)

## Desktop
- **OS:** OpenSUSE Tumbleweed
- **OS Version:** 2023xxxx
- **FreeTube version:** 0.19.0

## Additional context
- Opening 50 videos in YouTube with a single button click can be a pretty miserable experience for some browser + device combinations. With enough videos selected, it also enables a baby DoS attack on Invidious servers with a very low barrier to entry. This is clearly a childish and subpar attack vector which most/all Invidious servers have existing protections against, but making bad stuff easily accessible is a recipe for disaster. Instituting an arbitrary cap for this goes against the spirit of letting our most vocal [power]users do whatever they want, but 99% of users will probably not want to do that 99% of the time. After considering these concerns, I instituted an arbitrary count of 15 max videos in the selection that can be opened in one go, which is a fine place for now.
- Note that the icon title currently clearly indicates that this is a Video Selection Mode. This will not always be the case, but I made the label this way so as to best communicate its current limitations (i.e., only video entries, not playlists or channels).
- Saving and unsaving are notably the slowest operations when done to a large quantity of videos. Their performance could be improved by refactoring the save/unsave logic and updating it to use the batch addVideos call, but this is a lower priority with the Playlists PR on the horizon. I'd prefer that we wait until the Playlists PR is merged before making more sizable alterations in this section of the codebase.
- After this is merged, there will be a bit of work resolving the conflicts on the Playlists PR. I can take care of this, but it may not be immediate.
